### PR TITLE
Support multi-command sql statements

### PR DESCRIPTION
**Pull request by mckennagene**
split the input sql statement on ";" and execute each statement in order.

### CODE REVIEW

File: bin/tdsql [hunk 1]
```diff
@@ -389,11 +389,23 @@ my $db = TDSQL->new($HOSTNAME, $USERNAME, $PASSWORD);
    $db->output($OUTPUT) if has_value($OUTPUT);
    $db->format($FORMAT) if has_value($FORMAT);
 if($SQL) {
-    eval { $db->run_query($SQL, @ARGS) };
-    if(my $errstr = $@) {
-        $db->finish;
-        chomp $errstr;
-        die "$ME: $errstr\n";
+    my $counter = 1 ;
```
Comment (dbushong) [inline bin/tdsql]: not being used

File: bin/tdsql [hunk 2]
```diff
@@ -389,11 +389,23 @@ my $db = TDSQL->new($HOSTNAME, $USERNAME, $PASSWORD);
...
+    my @sql_statements = split(/;/,$SQL);
+    foreach my $statement(@sql_statements) {
+      $statement =~ s/^\s*//g;
+      $statement =~ s/\s*$//g;
```
Comment (dbushong) [inline bin/tdsql]: simpler:
```perl
my @sql_statements = split(/\s*;\s*/, $SQL);
for my $statement (@sql_statements) {
```

File: bin/tdsql [hunk 3]
```diff
@@ -389,11 +389,23 @@ my $db = TDSQL->new($HOSTNAME, $USERNAME, $PASSWORD);
...
+      if($statement ne "") {
+        # would like to have this output but it is being added
+        # to the output file if you to tdsql < file.sql > file.txt
+        # print STDERR "$counter/$#sql_statements ";
```
Comment (dbushong) [inline bin/tdsql]: understood; please put this as a PR comment rather than a code comment

File: bin/tdsql [hunk 4]
```diff
@@ -389,11 +389,23 @@ my $db = TDSQL->new($HOSTNAME, $USERNAME, $PASSWORD);
...
+        eval { $db->run_query($statement, @ARGS) };
+        if(my $errstr = $@) {
+          $db->finish;
+          chomp $errstr;
+          die "$ME: $errstr\n";
+        }
+      }
+      $counter++;
```
Comment (dbushong) [inline bin/tdsql]: not used

**Comment by andrewgho** [comment]
The reason `tdsql` didn't support running multiple statements from the start using semicolons as statement delimiters is that the approach of splitting on semicolon will fail if your SQL query has semicolons in it, for example:

```sql
SELECT text
  FROM quotes
 WHERE text LIKE '%With educated people, I suppose, punctuation is a matter of rule; with me it is a matter of feeling.%'
```

To properly support semicolons we would need an SQL parser, which seemed a bit heavyweight for this tool. Better yet would be for the database driver to support taking multiple statements and applying its own parsing.

**Comment by dbushong** [comment]
Good catch @andrewgho - I think you're right this will need a rethink

**Comment by mckennagene** [comment]
Thanks Andrew,

I'll change to make it an option, like -m for multi or something. I realize
it isn't perfect, but i think the use case you give happens extremely
rarely in real-world scenarios.

gene

**Comment by andrewgho** [comment]
I agree the use case of passing multiple SQL statements is a good one, and that having an embedded semicolon is rare. But it would be good to think of a way to do this that would be have well defined quoting semantics. I'm not attached to any of these, but here are some possible implementations that would have unambiguous semantics:

* Allow an optional `-e command` so that `tdsql -e 'SELECT * FROM ...'` is an alias for `tdsql 'SELECT * FROM ...'`. Allow multiple `-e` so that you can pass multiple SQL statements that way.
** Optionally, preserve the semantics of being able to pass placeholder variable values by requiring `-e` to be the final argument, so that you can pass: `tdsql -e sql1 var1 var2 -e sql2 var3 var4`
* Allow an optional count argument which is the number of SQL commands to read and execute (default is one). If placeholder variables are present, pass the same set of placeholder variable values to each SQL command invocation.
* Allow an optional flag which indicates that multiple arguments are multiple SQL commands, rather than one SQL command plus placeholder variable arguments. No placeholder variables can be used if this flag is present.

The first seems cleanest to me.

**Comment by dbushong** [comment]
Would just parsing enough SQL to recognize quoted semicolons be that hard?

**Comment by mckennagene** [comment]
i like the brainstorming. the basic use case for me is most often this

tdsql < file.sql > file.txt

the code could certainly distinguish between semi-colon inside single or
double quotes vs semi-colon not contained by quotes, a bit like how csv
parsers deal with embedded commas. more complicated regex but probably
manageable.

another idea, we could also use a different, multi-character delimiter, and
let you specify that, something like

tdsql -statement_delimiter="START_NEW_SQL" < file.sql > file.txt

then the author of the SQL can control how they write their file of course.
If I went this route I'd want to make sure the delimiter can be one that
starts with -- as well, so it could essentially be a comment in the SQL and
thus remain compatible with other SQL interpreters.

gene

**Comment by mckennagene** [comment]
submitted update, i forgot to comment in the update submission.

implemented -s (or --statement_delimiter) so user can specify a delimiter.
default behavior if user doesn't specify a delimiter is same as existing
tdql.

have tested with ; as delimiter and with -- next sql as delimiter.

gene

**Pull request closed without merging**